pageserver/src/thread_mgr.rs

```diff
@@ -0,0 +1,297 @@
+//!
+//! We provide a few basic facilities:
+//! - A global registry of threads that lists what kind of threads they are, and
+//!   which tenant or timeline they are working on
+//!
+//! - The ability request a thread to shut down.
+//!
+//! TODO: This would be a good place to also handle panics in a somewhat sane way
+//! Depending on what thread panics, we might want to kill the whole server, or
+//! only a single tenant or timeline.
+//!
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::panic;
+use std::panic::AssertUnwindSafe;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+
+use tokio::sync::watch;
+
+use tracing::{info, warn};
+
+use lazy_static::lazy_static;
+
+use zenith_utils::zid::{ZTenantId, ZTimelineId};
+
+lazy_static! {
+    /// Each thread that we track is associated with a "thread ID". It's just
+    /// an increasing number that we assign, not related to any system thread
+    /// id.
+    static ref NEXT_THREAD_ID: AtomicU64 = AtomicU64::new(1);
+
+    /// Global registry of threads
+    static ref THREADS: Mutex<HashMap<u64, Arc<PageServerThread>>> = Mutex::new(HashMap::new());
+}
+
+// There is a Tokio watch channel for each thread, which can be used to signal the
+// thread that it needs to shut down. This thread local variable holds the receiving
+// end of the channel. The sender is kept in the global registry, so that anyone
+// can send the signal to request thread shutdown.
+thread_local!(static SHUTDOWN_RX: RefCell<Option<watch::Receiver<()>>> = RefCell::new(None));
+
+// Each thread holds reference to its own PageServerThread here.
+thread_local!(static CURRENT_THREAD: RefCell<Option<Arc<PageServerThread>>> = RefCell::new(None));
+
+///
+/// There are many kinds of threads in the system. Some are associated with a particular
+/// tenant or timeline, while others are global.
+///
+/// Note that we don't try to limit how may threads of a certain kind can be running
+/// at the same time.
+///
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum ThreadKind {
+    // libpq listener thread. It just accepts connection and spawns a
+    // PageRequestHandler thread for each connection.
+    LibpqEndpointListener,
+
+    // HTTP endpoint listener.
+    HttpEndpointListener,
+
+    // Thread that handles a single connection. A PageRequestHandler thread
+    // starts detached from any particular tenant or timeline, but it can be
+    // associated with one later, after receiving a command from the client.
+    PageRequestHandler,
+
+    // Thread that connects to a safekeeper to fetch WAL for one timeline.
+    WalReceiver,
+
+    // Thread that handles checkpointing of all timelines for a tenant.
+    Checkpointer,
+
+    // Thread that handles GC of a tenant
+    GarbageCollector,
+
+    // Thread that checks status of remote storage. Shared by all tenants.
```

Comment:
```suggestion
    // Thread for synchronizing pageserver relish data with the remote storage. Shared by all tenants.
```

Also, as an idea to implement later:
> Shared by all tenants

could be enforced on a type level, so that we could not create such threads wrongfully.

Reply:
> Also, as an idea to implement later:
> 
> > Shared by all tenants
> 
> could be enforced on a type level, so that we could not create such threads wrongfully.

Yeah, I considered having 'timelineid' and 'tenantid' part of this Enum, for those thread kinds that are associated with timelines or tenants. But that's a bit strange for PageRequestHandlers which are initially not associated with timelines, but become associated later.

Now that I look more closely, I actually forgot to call `associate_thread` anywhere. So PageRequestHandlers are never associated with a timeline. That means that any threads that are currently handling requests for a timeline are not killed, when the timeline is shut down. Maybe that's OK and I was just trying to be too cute with this. I think I'll remove `associate_thread` since it's dead code, and leave that for a later patch if we want to do it.

Reply:
In general, I'd be very cautious when adding `pub` modifiers to such methods/functions, as it turns off rustc's warnings about unused.